storage/minifile/mimifile.go

```diff
@@ -0,0 +1,227 @@
+package minifile
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/prometheus/tsdb/fileutil"
+	"github.com/syndtr/goleveldb/leveldb/util"
+)
+
+type MiniFile struct {
+	path         string
+	instanceLock fileutil.Releaser // File-system lock to prevent double opens
+	lock         *sync.RWMutex
+	closed       int64
+}
+
+func New(path string) (*MiniFile, error) {
+	abs, err := filepath.Abs(path)
+	if err != nil {
+		return nil, err
+	}
+
+	flock, _, err := fileutil.Flock(filepath.Join(abs, "FLOCK"))
+	if err != nil {
+		return nil, err
+	}
+
+	err = os.MkdirAll(abs, 0755)
+	if err != nil {
+		return nil, err
+	}
+	return &MiniFile{
+		path:         abs,
+		instanceLock: flock,
+		lock:         &sync.RWMutex{},
+	}, nil
+}
+
+func (mf *MiniFile) Put(key string, value []byte) error {
+	if mf.isClosed() {
+		return fmt.Errorf("the miniFile storage is closed")
+	}
+
+	if key == "" {
+		return fmt.Errorf("store file with empty key")
+	}
+
+	crc := make([]byte, 4)
+	binary.LittleEndian.PutUint32(crc, util.NewCRC(value).Value())
+	value = append(value, crc...)
+
+	mf.lock.Lock()
+	defer mf.lock.Unlock()
+
+	name := filepath.Join(mf.path, key)
+
+	if err := ioutil.WriteFile(name, value, 0644); err != nil {
+		return fmt.Errorf("fail to write file %s: %w", name, err)
+	}
+
+	return nil
+}
+
+func (mf *MiniFile) Delete(key string) error {
+	if mf.isClosed() {
+		return fmt.Errorf("the miniFile storage is closed")
+	}
+
+	mf.lock.Lock()
+	defer mf.lock.Unlock()
+
+	err := os.Remove(filepath.Join(mf.path, key))
+	if err != nil && isNoFileError(err) {
+		return nil
+	}
+
+	return err
+}
+func (mf *MiniFile) Get(key string) ([]byte, error) {
+	if mf.isClosed() {
+		return nil, fmt.Errorf("the miniFile storage is closed")
+	}
+
+	mf.lock.RLock()
+
+	val, err := mf.get(key)
+	if err != nil {
+		mf.lock.RUnlock()
+		_ = mf.Delete(key)
```

Comment:
mf.Delete(key) change to os.Remove(path), and then release the RUnlock by defer ？

Reply:
fixed